utils/env_preferences/src/apple.rs

```diff
@@ -0,0 +1,167 @@
+    // copying it using `CFStringGetCString`
+    let lang_str: *const c_char = unsafe { CFStringGetCStringPtr(ptr, kCFStringEncodingUTF8) };
+
+    if !lang_str.is_null() {
+        // SAFETY: A valid `NULL` terminator is present which is a requirement of `from_ptr`
+        let lang_rust_str = unsafe { CStr::from_ptr(lang_str) }.to_str()?;
+        Ok(lang_rust_str.to_string())
+    } else {
+        // `lang_str` is null, i.e. `CFStringGetCStringPtr` couldn't give desired output, trying with
+        // manual allocations
+        // SAFETY: It returns length of the string, from above conditional statement we ensure
+        // that the `lang_ptr` is not NULL thus making it safe to call
+        let length = unsafe { CFStringGetLength(ptr) as usize };
+
+        let mut c_str_buf: Vec<u8> = vec![0; length + 1];
+
+        // SAFETY: Safety is ensured by following points
+        // 1. `lang_ptr` is not NULL, checked through conditional statement
+        // 2. `c_str_buf` is large enough and in scope after this call
+        unsafe {
+            CFStringGetCString(
+                ptr,
+                c_str_buf.as_mut_ptr() as *mut c_char,
+                c_str_buf.len() as CFIndex,
+                kCFStringEncodingUTF8,
+            );
+        }
+
+        let c_string = CString::from_vec_with_nul(c_str_buf)?;
+        let str_converted = c_string.into_string()?;
+        Ok(str_converted
+            .to_string()
+            .chars()
+            .filter(|&c| c != '\0')
+            .collect())
+    }
+}
+
+pub fn get_locales() -> Result<Vec<String>, RetrievalError> {
+    let mut languages: Vec<String> = Vec::new();
+
+    // SAFETY: The call to `CFLocaleCopyPreferredLanguages` returns an immutable reference to `CFArray` which is owned by us
+    // https://developer.apple.com/documentation/corefoundation/cfarrayref. It is ensured that `locale_carr_ref` is not mutated
+    // Immutablility ensures that nothing is overriden during it's scope
+    let locale_carr_ref = unsafe { CFLocaleCopyPreferredLanguages() };
+
+    if !locale_carr_ref.is_null() {
+        // SAFETY: The call to `CFArrayGetCount` is only made when is `locale_carr_ref` is not `NULL`
+        let count = unsafe { CFArrayGetCount(locale_carr_ref as _) };
+
+        for i in 0..count {
+            // SAFETY: The call to `CFArrayGetValueAtIndex` is safe because we are iterating from 0 to count (`CFArrayGetCount`) which
+            // in itself will always be greater than 0 and less than count ensuring we will not get "out of bounds" error
+            let lang_ptr = unsafe { CFArrayGetValueAtIndex(locale_carr_ref, i) };
+
+            if !lang_ptr.is_null() {
+                let locale_str = get_string(lang_ptr as CFStringRef)?;
+                languages.push(locale_str);
+            } else {
+                return Err(RetrievalError::NullPointer);
```

Comment:
why is this returning `NullPointer` instead of `NullLocale`?

Reply:
Because it's a pointer to the stored locale, the reference to `CFString`

Reply:
Isn't the case where you return `NullLocale` just that, a null string?

Reply:
Ok, if I think `NullLocale` is more descriptive as, if the we are getting a null it implies we couldn't the locale from the array

Reply:
You should now be able to get rid of the `NullPointer` variant.